src/Orleans.Runtime/MembershipService/LocalSiloHealthMonitor.cs

```diff
@@ -0,0 +1,405 @@
+        private readonly ILocalSiloDetails _localSiloDetails;
+        private readonly ILogger<LocalSiloHealthMonitor> _log;
+        private readonly ClusterMembershipOptions _clusterMembershipOptions;
+        private readonly IAsyncTimer _degradationCheckTimer;
+        private readonly ThreadPoolMonitor _threadPoolMonitor;
+        private ValueStopwatch _runTime;
+        private Task _runTask;
+        private DateTime _lastHealthCheckTime;
+
+        public LocalSiloHealthMonitor(
+            IEnumerable<IHealthCheckParticipant> healthCheckParticipants,
+            MembershipTableManager membershipTableManager,
+            ConnectionManager connectionManager,
+            ClusterHealthMonitor clusterHealthMonitor,
+            ILocalSiloDetails localSiloDetails,
+            ILogger<LocalSiloHealthMonitor> log,
+            IOptions<ClusterMembershipOptions> clusterMembershipOptions,
+            IAsyncTimerFactory timerFactory,
+            ILoggerFactory loggerFactory)
+        {
+            _healthCheckParticipants = healthCheckParticipants.ToList();
+            _membershipTableManager = membershipTableManager;
+            _connectionManager = connectionManager;
+            _clusterHealthMonitor = clusterHealthMonitor;
+            _localSiloDetails = localSiloDetails;
+            _log = log;
+            _clusterMembershipOptions = clusterMembershipOptions.Value;
+            _degradationCheckTimer = timerFactory.Create(
+                _clusterMembershipOptions.LocalHealthDegradationMonitoringPeriod,
+                nameof(LocalSiloHealthMonitor));
+            _threadPoolMonitor = new ThreadPoolMonitor(loggerFactory.CreateLogger<ThreadPoolMonitor>());
+            _runTime = ValueStopwatch.StartNew();
+        }
+
+        /// <summary>
+        /// Returns the local health degradation score, which is a value between 0 (healthy) and <see cref="MaxScore"/> (unhealthy).
+        /// </summary>
+        /// <param name="checkTime">The time which the check is taking place.</param>
+        /// <returns>The local health degradation score, which is a value between 0 (healthy) and <see cref="MaxScore"/> (unhealthy).</returns>
+        public int GetLocalHealthDegradationScore(DateTime checkTime)
+        {
+            var score = 0;
+            score += CheckSuspectingNodes(checkTime);
+            score += CheckReceivedProbeResponses(checkTime);
+            score += CheckReceivedProbeRequests(checkTime);
+            score += CheckLocalHealthCheckParticipants(checkTime);
+            score += CheckThreadPoolQueueDelay(checkTime);
+
+            // Clamp the score between 0 and the maximum allowed score.
+            score = Math.Max(0, Math.Min(MaxScore, score));
+            return score;
+        }
+
+        private int CheckThreadPoolQueueDelay(DateTime checkTime)
+        {
+            var threadPoolDelaySeconds = _threadPoolMonitor.MeasureQueueDelay().TotalSeconds;
+
+            if (threadPoolDelaySeconds > 1)
```

Comment:
>1 [](start = 41, length = 1)

nit, Could be a static reaonly Timespan.FromSeconds(1) so you didn't have to make the .TotalSeconds calculation above and could do direct TimeSpan comparison.

Reply:
The TotalSeconds becomes the addition to the score, so if the ThreadPool is delayed by 2 seconds, the score is incremented by 2, and so on, up to the max value of 8

Reply:
Ah, I see now. Makes sense. However, I don't see the max being enforced.

---
In reply to: [495134118](https://github.com/dotnet/orleans/pull/6745#discussion_r495134118) [](ancestors = 495134118)

Reply:
It's in the top-level GetLocalHealthDegradationScore method, at the end: https://github.com/dotnet/orleans/pull/6745/files#diff-c163052ab14eb8bf8ec18a979094576cR94